File: godfather/roles/mixins/shooter.py
from godfather.game.types import Attack, Defense, Priority


class Shooter:
    def __init__(self, *args, **kwargs):
        self.action = 'shoot'
        self.action_gerund = 'shooting'
        self.action_priority = Priority.SHOOTER
        self.action_text = 'shoot a player'
        super().__init__(*args, **kwargs)

    async def run_action(self, actions, player, target):
        if hasattr(player.role, 'bullets'):
            player.role.bullets -= 1
        if target.role.defense() > Defense.NONE:
            return
        pl_record = actions.record[target.user.id]
        pl_record['nightkill']['result'] = True
        pl_record['nightkill']['type'] = Attack.BASIC
        pl_record['nightkill']['by'].append(player)

File: src/lib/mafia/mixins/Shooter.ts
import NightActionsManager, { Attack, NightActionPriority } from '@mafia/managers/NightActionsManager';
import Player from '@mafia/Player';
import ActionRole from '@mafia/mixins/ActionRole';

export default class Shooter extends ActionRole {

	public action = 'shoot';
	public actionText = 'shoot a player';
	public actionGerund = 'shooting';
	public priority = NightActionPriority.Shooter;
	public bullets = Infinity;
	public flags = {
		canBlock: true,
		canTransport: true,
		canVisit: true
	};

	public runAction(actions: NightActionsManager, target: Player) {
		this.bullets--;
		if (target.role.defense > this.attackStrength) return;
		actions.record.setAction(target.user.id, 'nightkill', { result: true, by: [this.player], type: Attack.Basic });
	}

	public get attackStrength() {
		return Attack.Basic;
	}

}
